### PR TITLE
Added e2e test for the word-count example

### PR DESCRIPTION
**Pull request by cddr**
Demonstrates how to use the test-machine to write a test that can be executed against the mock, kafka, or rest-proxy transports depending on the environment variables present.

Testing:
```
cd examples/word-count
# look ma! no kafka
clj -A:test --namespace word-count-e2e-test

# test against local kafka
docker-compose up -d
BOOTSTRAP_SERVERS=localhost:9092 clj -A:test --namespace word-count-e2e-test

# test against "remote" kafka
# start the word-count app running against the cluster setup by docker-compose (relies on topics created by running previous tests)
docker run --network word-count_default \
  -v $(pwd):/usr/src/app \
  -v $HOME/.m2:/root/.m2 \
  -w /usr/src/app \
  -e BOOTSTRAP_SERVERS=broker:9092 \
  clojure:tools-deps clj -m word-count

# run the tests against the rest-proxy setup by docker-compose (no direct access required)
REST_PROXY_URL=http://localhost:8082 clj -A:test --namespace word-count-e2e-test
```

### CODE REVIEW

**Comment by funderbar** [comment]
When I was first trying out jackdaw I tried to figure out how to use examples easily, but there seems to be a bar for that and the description in READ.ME does not really work. 

What is the usual process for it

**Comment by creese** [comment]
@funderbar Did you follow the directions in the comment block?
https://github.com/FundingCircle/jackdaw/blob/master/examples/word-count/src/word_count.clj#L106-L286

File: src/jackdaw/test/fixtures.clj [hunk 1]
```diff
@@ -162,6 +164,71 @@
...
+                                "/tmp/kafka-streams")
+                            (get app-config "application.id"))]
+      (when (.exists (io/file state-dir))
+        (log/info "deleting state dir: " state-dir)
+        (delete-recursively state-dir))
+      (t))))
+
+;;; reset-application-fixture ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn reset-application-fixture [app-config]
```
Comment (creese) [inline src/jackdaw/test/fixtures.clj]: In the reloaded workflow, Stuart Sierra defines "reset" this way:
> Finally, the heart of my workflow: the reset function. This is one function which I can call at the REPL to 1) stop the current application instance; 2) reload any source files that have changed; and 3) create and start a new application instance.

In `user/reset`, we don't reload source files, but we do stop the app, clean up, create topics, and restart. This seems closer to the `integration-fixture` fixture than the `reset-application-fixture`. How should we align the two approaches?

File: examples/word-count/test/word_count_e2e_test.clj [hunk 2]
```diff
@@ -0,0 +1,111 @@
+(ns word-count-e2e-test
```
Comment (creese) [inline examples/word-count/test/word_count_e2e_test.clj]: I'd be okay with replacing the other test namespace with this one.

File: examples/word-count/test/word_count_e2e_test.clj [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      (fn [machine]
+        (let [lines ["As Gregor Samsa awoke one morning from uneasy dreams"
+                     "he found himself transformed in his bed into an enormous insect"
+                     "What a fate: to be condemned to work for a firm where the"
+                     "slightest negligence at once gave rise to the gravest suspicion"
+                     "How about if I sleep a little bit longer and forget all this nonsense"
+                     "I cannot make you understand"]
+              commands (->> (concat
+                             (map input-writer lines)
+                             [(word-watcher "understand")]))
```
Comment (creese) [inline examples/word-count/test/word_count_e2e_test.clj]: Can the commands be written in pure EDN? Is it worth doing that?

File: src/jackdaw/test/fixtures.clj (hunk 1 above)
Comment (cddr) [reply src/jackdaw/test/fixtures.clj]: The "reset" terminology here has nothing to do with re-loaded. It's the [application reset tool](https://docs.confluent.io/current/streams/developer-guide/app-reset-tool.html) from kafka

File: examples/word-count/test/word_count_e2e_test.clj (hunk 3 above)
Comment (cddr) [reply examples/word-count/test/word_count_e2e_test.clj]: It was like that in the first iteration. This seemed clearer IMO.

File: examples/word-count/test/word_count_e2e_test.clj [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+       last
+       :value))
+
+(defn get-env [k]
+  (get (System/getenv) k))
+
+(def test-config
+  {:broker-config broker-config
+   :topic-config wc/topic-metadata
+   :kstream-config (assoc wc/app-config "cache.max.bytes.buffering" "0")
```
[the remark this answers is not in the document]
Comment (cddr) [reply examples/word-count/test/word_count_e2e_test.clj]: Having it here will mean it continues to behave more predictably in a test context even when the application is fixed. With "cache.max.bytes.buffering=0" is likely to induce high write throughput.

https://kafka.apache.org/10/documentation/streams/developer-guide/memory-mgmt.html

File: src/jackdaw/test/fixtures.clj (hunk 1 above)
Comment (creese) [reply src/jackdaw/test/fixtures.clj]: Let's hold this until the next PR as discussed.

File: examples/word-count/test/word_count_e2e_test.clj (hunk 4 above)
[the remark this answers is not in the document]
Comment (creese) [reply examples/word-count/test/word_count_e2e_test.clj]: > 2- `wc/app-config` already contains `"cache.max.bytes.buffering" "0"`. There is no need to assoc.
I'm resolving this one.

File: examples/word-count/test/word_count_e2e_test.clj [hunk 5]
```diff
@@ -0,0 +1,111 @@
+(ns word-count-e2e-test
+  (:require
+   [word-count :as wc]
+   [jackdaw.streams :as k]
+   [jackdaw.streams.protocols :as proto]
+   [jackdaw.test :as jd.test]
```
[the remark this answers is not in the document]
Comment (creese) [reply examples/word-count/test/word_count_e2e_test.clj]: Let's save this for a future discussion.

File: examples/word-count/test/word_count_e2e_test.clj [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  [:write! :input line {:key-fn identity}])
+
+(defn word-watcher
+  "Builds a test-command that blocks until the supplied word appears"
+  [word]
+  [:watch (fn [journal]
+            (some #(= word (:key %))
+                  (get-in journal [:topics :output]))) 2000])
+
+(defn wc
```
[the remark this answers is not in the document]
Comment (creese) [reply examples/word-count/test/word_count_e2e_test.clj]: Now that I understand this a bit better, I thing should just spell it out, i.e. `word-count`.

**Pull request merged**